x-pack/plugins/code/server/repository_service.ts

```diff
@@ -19,10 +21,24 @@ import {
 } from '../model';
 import { Logger } from './log';
 
-export type CloneProgressHandler = (progress: number, cloneProgress?: CloneProgress) => void;
+// Return false to stop the clone progress. Return true to keep going;
+export type CloneProgressHandler = (progress: number, cloneProgress?: CloneProgress) => boolean;
+export type UpdateProgressHandler = () => boolean;
 
+const GIT_FETCH_PROGRESS_CANCEL = -1;
+// TODO: Cannot directly access Git.Error.CODE.EUSER (-7). Investigate why.
+const NODEGIT_CALLBACK_RETURN_VALUE_ERROR = -7;
+const GIT_INDEXER_PROGRESS_CALLBACK_RETURN_VALUE_ERROR_MSG = `indexer progress callback returned ${GIT_FETCH_PROGRESS_CANCEL}`;
 const SSH_AUTH_ERROR = new Error('Failed to authenticate SSH session');
 
+function isCancelled(error: any) {
+  return (
+    error &&
+    (error.message.includes(GIT_INDEXER_PROGRESS_CALLBACK_RETURN_VALUE_ERROR_MSG) ||
+      error.errno === NODEGIT_CALLBACK_RETURN_VALUE_ERROR)
```

Comment:
why we returned  `GIT_FETCH_PROGRESS_CANCEL`, but catch `NODEGIT_CALLBACK_RETURN_VALUE_ERROR`?

Reply:
the returned value won't be populated in this case. it will always return `-7`. but for no reason, you can not import the constant from the nodegit, otherwise, it always throws errors in runtime. so add our own constant here.

in short, canceling the clone/update job in this way will trigger either case of these 2 here. So wrap it in a function in case any further changes.